### PR TITLE
8329998: Remove double initialization for parts of small TypeArrays in ZObjArrayAllocator

### PR DESCRIPTION
**Pull request by xmas92**
[JDK-8139457](https://bugs.openjdk.org/browse/JDK-8139457) introduced non-word aligned payloads for TypeArrays. This required adaptation in `ZObjArrayAllocator` segmented array allocator. But the implementation does this adaptation even when the size is small enough and dispatches the allocation to `ObjArrayAllocator`, which already handles clearing this memory.

Instead move the initialization of any unaligned payload start to after the segmented allocation decision.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329998](https://bugs.openjdk.org/browse/JDK-8329998): Remove double initialization for parts of small TypeArrays in ZObjArrayAllocator (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18721/head:pull/18721` \
`$ git checkout pull/18721`

Update a local copy of the PR: \
`$ git checkout pull/18721` \
`$ git pull https://git.openjdk.org/jdk.git pull/18721/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18721`

View PR using the GUI difftool: \
`$ git pr show -t 18721`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18721.diff">https://git.openjdk.org/jdk/pull/18721.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18721#issuecomment-2047507969)

### CODE REVIEW

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/gc/z/zObjArrayAllocator.cpp [hunk 1]
```diff
@@ -88,6 +75,22 @@ oop ZObjArrayAllocator::initialize(HeapWord* mem) const {
...
 
+  const BasicType element_type = ArrayKlass::cast(_klass)->element_type();
+  const size_t base_offset_in_bytes = arrayOopDesc::base_offset_in_bytes(element_type);
+  const size_t process_start_offset_in_bytes = align_up(base_offset_in_bytes, BytesPerWord);
+
+  if (process_start_offset_in_bytes != base_offset_in_bytes) {
+    // initialize_memory can only fill word aligned memory,
+    // fill the first 4 bytes here.
+    assert(process_start_offset_in_bytes - base_offset_in_bytes == 4, "Must be 4-byte aligned");
+    assert(!is_reference_type(element_type), "Only TypeArrays can be 4-byte aligned");
```
Comment (lgxbslgx) [inline src/hotspot/share/gc/z/zObjArrayAllocator.cpp:86]: Actually, only the type arrays whose element type is not double word can be 4-byte aligned. So this assertion could be `!is_reference_type(element_type) && !is_double_word_type(element_type)`.

**Review by lgxbslgx — APPROVED** [review]
Looks good. One suggestion.

File: src/hotspot/share/gc/z/zObjArrayAllocator.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/gc/z/zObjArrayAllocator.cpp:86]: I think the point is that when it is not a reference type, it is OK to initialize with a plain `0`.

Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zObjArrayAllocator.cpp:86]: May be better to treat the assertion as a more detailed guard or check. But I'm also OK with the current code.

**Comment by xmas92** [comment]
Thanks for the reviews. Reran testing after merge.
/integrate

**Pull request closed without merging**